--- a/ggml-cuda/ggml-cuda.cu
+++ b/ggml-cuda/ggml-cuda.cu
@@ -785,7 +785,7 @@
 
     ggml_backend_cuda_buffer_context * ctx = new ggml_backend_cuda_buffer_context(buft_ctx->device, dev_ptr);
 
-    return ggml_backend_buffer_init(buft, ggml_backend_cuda_buffer_interface, ctx, size);
+    return ggml_backend_buffer_init(buft, &ggml_backend_cuda_buffer_interface, ctx, size);
 }
 
 static size_t ggml_backend_cuda_buffer_type_get_alignment(ggml_backend_buffer_type_t buft) {
